### PR TITLE
Improve Travis matrix

### PR DESCRIPTION
**Pull request by snapshotpl**
* Improve deps testing,
* add php 7.1,
* php7.* must pass tests

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,21 +1,43 @@
...
-  - 5.6
-  - 5.5
-  - 5.4
-  - hhvm
-  - hhvm-nightly
-
 matrix:
+  include:
+    - php: 5.4
+      env: DEPS=lowest
```
Comment (theofidry) [inline .travis.yml]: `COMPOSER_FLAGS='--prefer-lowest'`

File: .travis.yml [hunk 2]
```diff
@@ -1,21 +1,43 @@
...
+    - php: 5.4
```
Comment (theofidry) [inline .travis.yml]: I think you can skip the flags here as it's the default behaviour

File: .travis.yml [hunk 3]
```diff
@@ -1,21 +1,43 @@
...
+      env: DEPS=latest
+    - php: 5.5
+      env: DEPS=lowest
+    - php: 5.5
+      env: DEPS=latest
```
Comment (theofidry) [inline .travis.yml]: doubling the number of tests looks a bit too much to me. Having a test for the two extrema plus one for each version is enough I think

File: .travis.yml [hunk 4]
```diff
@@ -1,21 +1,43 @@
...
+      env: DEPS=lowest
+    - php: 5.6
+      env: DEPS=latest
+    - php: 7.0
+      env: DEPS=lowest
+    - php: 7.0
+      env: DEPS=latest
+    - php: 7.1
+      env: DEPS=lowest
+    - php: 7.1
```
Comment (theofidry) [inline .travis.yml]: we're missing a `nightly` build that should be allowed to fail

File: .travis.yml [hunk 5]
```diff
@@ -28,8 +50,9 @@ before_install:
   - travis_retry composer self-update
```
Comment (theofidry) [inline .travis.yml]: this is no longer necessary, although it's not related to this PR I think you can take the opportunity to remove it

File: .travis.yml [hunk 6]
```diff
@@ -28,8 +50,9 @@ before_install:
...
 
 before_script:
-  - travis_retry composer install --no-interaction
+  - if [[ $DEPS == 'latest' ]]; then travis_retry composer update --no-interaction --prefer-source ; fi
```
Comment (theofidry) [inline .travis.yml]: With my first comment, this [the two added lines] can be simplified to `composer update --no-interaction --prefer-source $COMPOSER_FLAGS`

File: .travis.yml [hunk 7]
```diff
@@ -28,8 +50,9 @@ before_install:
...
+  - if [[ $DEPS == 'lowest' ]]; then travis_retry composer update --no-interaction --prefer-source --prefer-lowest ; fi
 
 script:
-  - vendor/bin/phpunit --coverage-text
+  - if [[ $TEST_COVERAGE == 'true' ]]; then vendor/bin/phpunit --coverage-text ; else vendor/bin/phpunit ; fi
```
Comment (theofidry) [inline .travis.yml]: `if [ -n "$TEST_COVERAGE" ]; then`?

Comment (theofidry) [inline .travis.yml]: Single brackets are also enough AFAIK

File: .travis.yml [hunk 8]
```diff
@@ -1,21 +1,43 @@
 language: php
 sudo: false
```
Comment (theofidry) [inline .travis.yml]: also no part of the goal of this PR, can be removed as well

**Comment by piotr-zuralski** [comment]
Could you please adapt also those changes in our PR:

composer.json
````
"phpunit/phpunit": "^4.5|^5.0|<=6.0"
````
new PHPUnit 6.0 does not support tap parameter, TAP format is not supported any more - https://github.com/sebastianbergmann/phpunit/blob/master/ChangeLog-6.0.md#removed

features/version.feature
replace `Humbug version 1.0-dev` with `Humbug 1.0-dev` current Humbug ` humbug.phar --version` also doesn't print `Humbug version (...)`

**Comment by snapshotpl** [comment]
@piotr-zuralski I think this changes what you said should be part of other pull requrest.

**Comment by piotr-zuralski** [comment]
@snapshotpl Those changes are already waiting in other PRs and without them builds will keep failing

**Comment by snapshotpl** [comment]
Ok, so when they will be merged into develop/master I can rebase it here. There's no sense to duplicate work

File: .travis.yml [hunk 9]
```diff
@@ -1,21 +1,32 @@
...
+      env:
+        - TEST_COVERAGE=true
+    - php: hhvm
+      env: COMPOSER_FLAGS='--prefer-lowest'
+    - php: hhvm
   fast_finish: true
   allow_failures:
-    - php: 7.0
     - php: hhvm
     - php: hhvm-nightly
```
Comment (pamil) [inline .travis.yml]: There's no `hhvm-nightly` build now :)

File: .travis.yml [hunk 10]
```diff
@@ -24,12 +35,9 @@ cache:
   directories:
     - $HOME/.composer/cache
 
-before_install:
-  - travis_retry composer self-update
-
 before_script:
-  - travis_retry composer install --no-interaction
+  - composer update --no-interaction --prefer-source $COMPOSER_FLAGS
```
Comment (pamil) [inline .travis.yml]: Is there some reason to put `--prefer-source` here? Installing source packages takes more time than dist ones.

File: .travis.yml [hunk 11]
```diff
@@ -24,12 +35,9 @@ cache:
...
 
 script:
-  - vendor/bin/phpunit --coverage-text
+  - if [[ $TEST_COVERAGE == 'true' ]]; then vendor/bin/phpunit --coverage-text ; else vendor/bin/phpunit ; fi
```
Comment (pamil) [inline .travis.yml]: What do you think about introducing `$PHPUNIT_FLAGS`? This way we can remove `if` from this statement.

**Comment by pamil** [comment]
Closing and reopening to force build on `humbug/humbug` :)

**Comment by snapshotpl** [comment]
Give 10 minutes to fix that

File: .travis.yml [hunk 12]
```diff
@@ -1,31 +1,37 @@
...
-  - 5.6
-  - 5.5
-  - 5.4
-  - hhvm
-  - hhvm-nightly
 
 matrix:
+  include:
+    - php: 5.4
+      env: COMPOSER_FLAGS='--prefer-lowest'
```
Comment (padraic) [inline .travis.yml]: Would it be simpler to just add an env block later? e.g.
```
env:
  matrix:
    - COMPOSER_FLAGS="--prefer-lowest"
    - COMPOSER_FLAGS=""
```
Keeps the list of flags and versions distinct?

File: .travis.yml [hunk 13]
```diff
@@ -1,31 +1,37 @@
...
-  - 7.0
-  - 5.6
-  - 5.5
-  - 5.4
-  - hhvm
-  - hhvm-nightly
 
 matrix:
+  include:
+    - php: 5.4
```
Comment (padraic) [inline .travis.yml]: Per #226, we can remove the EOL php versions 5.3 to 5.5.

File: .travis.yml [hunk 14]
```diff
@@ -1,27 +1,27 @@
...
-
 before_script:
-  - travis_retry composer install --no-interaction
+  - composer update --no-interaction --prefer-dist $COMPOSER_FLAGS
 
 script:
-  - vendor/bin/phpunit --coverage-text
-  - vendor/bin/behat
+  - vendor/bin/phpunit $PHPUNIT_FLAGS
+  - vendor/bin/behat -vv
```
Comment (theofidry) [inline .travis.yml]: I don't think we should put it in verbose mode, however we could do `-f progress --stop-on-failure` and add a `--stop-on-failure` for PHPUnit as well

File: .travis.yml [hunk 15]
```diff
@@ -1,27 +1,27 @@
...
 cache:
   directories:
     - $HOME/.composer/cache
 
-before_install:
-  - travis_retry composer self-update
-
 before_script:
-  - travis_retry composer install --no-interaction
+  - composer update --no-interaction --prefer-dist $COMPOSER_FLAGS
```
Comment (theofidry) [inline .travis.yml]: I would add `--no-suggest` as well

File: .travis.yml [hunk 16]
```diff
@@ -1,27 +1,27 @@
...
-sudo: false
-
-php:
-  - 5.6
-  - 7.0
-  - hhvm
-
 matrix:
+  include:
+    - php: 5.6
```
Comment (theofidry) [inline .travis.yml]: versions should be strings

File: .travis.yml [hunk 17]
```diff
@@ -1,27 +1,27 @@
...
+      env: PHPUNIT_FLAGS='--coverage-text'
+    - php: hhvm
   fast_finish: true
   allow_failures:
-    - php: 7.0
     - php: hhvm
 
+env:
+  matrix:
+    - COMPOSER_FLAGS="--prefer-lowest"
```
Comment (theofidry) [inline .travis.yml]: to stay consistent let's keep it at single quotes :)

File: .travis.yml (hunk 15 above)
Comment (snapshotpl) [inline .travis.yml]: Make a sense

File: .travis.yml (hunk 14 above)
Comment (snapshotpl) [inline .travis.yml]: `-f progress` - good idea. Verbose can be helpful for debugging without running on local machine.
`--stop-on-failure`? I think is better to get all issues in one build.

**Comment by snapshotpl** [comment]
This PR shows that humbug's tests don't work well with phpunit in version 5.0. Should we remove phpunit >5.0 features or change minimum version of phpunit?

**Comment by padraic** [comment]
Yes, there's been a lot of drift over time so it's messy getting the CI just right. The basic migration plan is to do just enough to make Humbug pass in its present state. Once "just enough" is accomplished, we're going straight to 2.0 on Humbug, and we can support PHPUnit 5-6 any way we want.

That may require paring back on PHPUnit features for a brief period (e.g. using exception annotations, not the methods, and so on, which are the same across 4,5&6) - also we have hard upper limits (Humbug 1 needs TAP support, for e.g.).

File: .travis.yml (hunk 14 above)
Comment (theofidry) [inline .travis.yml]: both non verbose + progress + fail fast are for fast builds detecting as early as possible if things pass or not. Figuring out why should be done locally, and for issues where it's hard to reproduce locally the user can always change the verbosity.

**Comment by snapshotpl** [comment]
PHPUnit compatibility improved. However php 7 still fall. But it's problem for another PR, because this scenario alread falls in master https://travis-ci.org/humbug/humbug/jobs/223862748 . disable `allow_failures` change game rules ;)

**Comment by padraic** [comment]
Thanks @snapshotpl!

**Comment by theofidry** [comment]
I'm really missing the thanks in the PRs lately... soz about that

**Pull request merged**